Adding Flatpak module

JayKayy opened the pull request:
##### SUMMARY
This change adds the `flatpak` module. This initial piece is a simple install/uninstall module for flatpaks and flatpak remotes.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
flatpak

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
$ ansible --version
ansible 2.4.0 (devel e3698d8d37) last updated 2017/07/26 15:26:31 (GMT -400)
  config file = /home/jkwiatko/.ansible.cfg
  configured module search path = [u'/home/jkwiatko/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jkwiatko/github/ansible/lib/ansible
  executable location = /home/jkwiatko/github/ansible/bin/ansible
  python version = 2.7.13 (default, Jun 26 2017, 10:20:05) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```
##### ADDITIONAL INFORMATION
example usage:
```
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://jgrulich.fedorapeople.org/telegram/telegram.flatpakref state=present'
localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.telegram.desktop/x86_64/master    system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://git.gnome.org/browse/gnome-apps-nightly/plain/gedit.flatpakref state=present'
localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.gnome.gedit/x86_64/master         system,current
org.telegram.desktop/x86_64/master    system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://git.gnome.org/browse/gnome-apps-nightly/plain/gedit.flatpakref state=absent'localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.telegram.desktop/x86_64/master    system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=https://jgrulich.fedorapeople.org/telegram/telegram.flatpakref state=absent'
localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current

jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
org.gnome.gedit/x86_64/master         system,current
[jkwiatko ansible]$ sudo ansible -M ~/.ansible/plugins/modules 'localhost' -m flatpak -a 'name=org.gnome.gedit state=absent'localhost | SUCCESS => {
    "changed": true
}
[jkwiatko ansible]$ flatpak list --app
Ref                                   Options       
com.spotify.Client/x86_64/master      system,current
com.valvesoftware.Steam/x86_64/stable system,current
[jkwiatko ansible]$ 

```

JayKayy committed:
Adding Flatpak module

JayKayy committed:
Adding Flatpak module

sysadmind commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,254 @@
+  name:
+    description:
+      -
+        "When state is set to `present`, `name` is best used as `http(s)` url format.
```
I believe that this section should start on the same line as the dash instead of the line beneath from looking at other modules.

sysadmind commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,254 @@
+  state:
+    description:
+      -
+        "Set to `present` will install the flatpak and/or remote.
```
This should also start on the line above if the other is moved.

sysadmind commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,254 @@
+module: flatpak
+version_added: "2.4"
+requirements:
+    - flatpak
```
The flatpak command is listed as a requirement but the module doesn't check if it is before proceeding. It would be a good idea to `module.fail_json()` with a useful message when that occurs. You can actually even get the full path with the function `module.get_bin_path()` as illustrated here: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/system/service.py#L451
That would also get you a full path to execute so that you don't run into relative path issues.

sysadmind reviewed (commented):
Overall looks pretty good. Just a few doc recommendations to follow other modules and some useful feedback to users on errors.

sysadmind commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+
+
+def install_flat(flat):
+    command = "flatpak install -y --from {}".format(flat)
```
Each of these functions creates a command but uses the relative executable `flatpak` but you have the absolute path below as `location` and you fail if it's not found. You could pass that location as an agrument or create a class wrapper around the functions. It would ensure that the initial flatpak path would always be used and possibly prevent environment issues.

andytom commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+        When set to `absent` the same `http(s)` will try to remove it using the
+        name of the flatpakref. However, there is no naming standard between
+        names of flatpakrefs and what the reverse DNS name of the installed flatpak
+        will be. Given that, it is best to use the http(s) url for name=`present`
```
Should this be ```... for state=`present` ``` ?

JayKayy replied:
You are correct on both of these accounts

andytom commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+        name of the flatpakref. However, there is no naming standard between
+        names of flatpakrefs and what the reverse DNS name of the installed flatpak
+        will be. Given that, it is best to use the http(s) url for name=`present`
+        and the reverse DNS name=`absent`. Alternatively reverse dns format can optimally
```
As above should this be ```...reverse DNS state=`absent` ...```?

andytom reviewed (commented):
Few minor changes to the docs.

mattclay committed:
Revert "Revert "Allow ini plugin to load file using other encoding than utf8." (#27407)"

This reverts commit 520696fb3949f708f55bce74ae2f2ab302fc9c21.

mattclay committed:
Revert "Revert "Allow ini plugin to load file using other encoding than utf8." (#27407)"

This reverts commit 520696fb3949f708f55bce74ae2f2ab302fc9c21.

Spredzy committed:
Removing openssl_publickey from pep8 legacy files (#27414)

sdodsley committed:
Add Pure Storage Volume module (#26357)

Update documentation to fix typos and define the extend feature
Force size to be a required variable on create as requested.
Change type to bool

trishnaguha committed:
fix iosxr_banner (#27378)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

sysadmind commented:
shipit

@JayKayy I'm really happy to see flatpak support land in Ansible. Nice work.

JayKayy commented:
Thanks. I certainly found it useful for my personal laptop setup :)

andytom reviewed (approved):
shipit

pmakowski commented:
shipit

jhawkesworth commented:
bot_status

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+requirements:
+    - flatpak
+author:
+    - "John Kwiatkoski (@jaykayy)"
```
This doesn't need to be quoted.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    - "John Kwiatkoski (@jaykayy)"
+short_description: "Install and remove flatpaks."
+description:
+    - "The flatpak module allows users to manage installation and removal of flatpaks."
```
This doesn't need to be quoted.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    - flatpak
+author:
+    - "John Kwiatkoski (@jaykayy)"
+short_description: "Install and remove flatpaks."
```
This doesn't need to be quoted.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    default: ''
+  remote:
+    description:
+      - "The flatpak remote repo to be used in the flatpak operation."
```
This doesn't need to be quoted.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+'''
+EXAMPLES = '''
+
+#install the spotify flatpak
```
Convert the task comments into the task names:

```
- name: Install the spotify flaptalk
...
```

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+ - flatpak:
+    name: org.gnome.gedit
+    state: absent
+
```
There should be no empty line here.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    name: org.gnome.gedit
+    state: absent
+
+'''
```
There should be new line behind this line.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    returned: always
+    type: string
+    sample: "https://sdk.gnome.org/gnome-apps.flatpakrepo"
+'''
```
There should be 2 new lines behind this line.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+        common_name = app
+    else:
+        common_name = name
+
```
Either remove this empty line OR add the same empty line into all function in front of the return command (preferred solution).

JayKayy replied:
Done

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+options:
+  name:
+    description:
+      - "When state is set to `present`, `name` is best used as `http(s)` url format.
```
You don't have to quote it even if the text spans multiple lines. Also, it should use the standard formatting for names and values of options and modules as described [here](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#formatting-options).

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    default: ''
+  state:
+    description:
+      - "Set to `present` will install the flatpak and/or remote.
```
The same like above.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    type: string
+    sample: "error while installing..."
+name:
+    description: "Name of flatpak given for the operation"
```
No quotes here.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    description: on failure, the output for the failure
+    returned: failed
+    type: string
+    sample: "error while installing..."
```
No quotes here.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+'''
+RETURN = '''
+reason:
+    description: on failure, the output for the failure
```
Capital letter in the value.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    description: "Name of flatpak given for the operation"
+    returned: always
+    type: string
+    sample: "https://git.gnome.org/.../gnome-apps/gedit.flatpakref"
```
No quotes here.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    type: string
+    sample: "https://git.gnome.org/.../gnome-apps/gedit.flatpakref"
+remote:
+    description: "remote of flatpak given for the operation"
```
No quotes here.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    description: "remote of flatpak given for the operation"
+    returned: always
+    type: string
+    sample: "https://sdk.gnome.org/gnome-apps.flatpakrepo"
```
No quotes here.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+        argument_spec=dict(
+            name=dict(required=False, default=''),
+            remote=dict(required=False, default=''),
+            state=dict(required=False, default="present",
```
This should be wrapped like this:

```
            state=dict(
                required=False, default="present", choices=['present', 'absent'])
```

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+    location = module.get_bin_path('flatpak')
+    if location is None:
+        module.fail_json(
+            msg="cannot find \'flatpak\' binary. Aborting.")
```
No backslashes here.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,256 @@
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+DOCUMENTATION = '''
```
There should be a new line above this line.

JayKayy committed:
Update flatpak.py

Feedback based changes

JayKayy committed:
Update flatpak.py

Feedback based changes

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,259 @@
+    default: present
+'''
+EXAMPLES = '''
+
```
You can delete these empty lines.

JayKayy committed:
Update flatpak.py

Adding newlines before return statements noted in the preferred solution.

removing empty lines in examples

JayKayy committed:
Update flatpak.py

Adding newlines before return statements noted in the preferred solution.

removing empty lines in examples

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,267 @@
+options:
+  name:
+    description:
+      - When I(state) is set to `C(present)`, `I(name)` is best used as `http(s)` url format.
```
You can remove the backticks when using the other decoration.

JayKayy commented:
I have messed up my branch running into merge conflicts with files i didnt modify. I will be fixing these shortly.

JayKayy committed:
resolving issue

JayKayy committed:
resolving issue

JayKayy committed:
fixing documentation

JayKayy committed:
fixing documentation

JayKayy committed:
fixing hanging indent

JayKayy committed:
fixing hanging indent

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,264 @@
+
+RETURN = '''
+reason:
+    description: On failure, the output for the failure
```
This should be indented by 2 spaces. The same bellow.

JayKayy replied:
done

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,264 @@
+options:
+  name:
+    description:
+      - When I(state) is set to C(present), I(name) is best used as `http(s)` url format.
```
You can drop the backticks around the `http(s)`.

JayKayy replied:
done

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,264 @@
+# GNU General Public License v3.0+ (see COPYING or
+# https://www.gnu.org/licenses/gpl-3.0.txt)
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
```
I would format this like this:

```
ANSIBLE_METADATA = {
    'metadata_version': '1.0',
    'status': ['preview'],
    'supported_by': 'community'
}
```

JayKayy replied:
done

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,264 @@
+def install_flat(binary, flat):
+    command = "{} install -y --from {}".format(binary, flat)
+    output = flatpak_command(command)
+    if 'error' in output and 'already installed' not in output:
```
What about to insert new line in front of the `if`? It would make nice visual separation between the variable definitions and the condition. The same at any place bellow.

JayKayy replied:
This makes sense to me.

JayKayy replied:
Done

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,264 @@
+requirements:
+    - flatpak
+author:
+    - John Kwiatkoski (@jaykayy)
```
This should be indented only by 2 spaces.

JayKayy replied:
done

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,264 @@
+    - John Kwiatkoski (@jaykayy)
+short_description: Install and remove flatpaks.
+description:
+    - The flatpak module allows users to manage installation and removal of flatpaks.
```
This should be indented only by 2 spaces.

JayKayy replied:
done

dagwieers replied:
Actually this shouldn't be indented at all. Lists can start from the same indentation as their parent.
And the general rule is, don't add things that aren't needed (so don't quote strings if they're not needed, or don't indent if it's not needed).

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,264 @@
+    required: false
+    default: present
+'''
+EXAMPLES = '''
```
I would insert one new line in front of this.

JayKayy replied:
done

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,274 @@
+ANSIBLE_METADATA = {
+    'metadata_version': '1.0',
+    'status': ['preview'],
+    'supported_by': 'community'}
```
I would put the final cruelly bracket on a new line:

```
    ...
    'supported_by': 'community'
}
```

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,274 @@
+                module.exit_json(changed=True)
+
+            code, output = add_remote(location, remote)
+            if code == 1:
```
What about to separate the `if` blocks here as well as you did above? The same bellow.

jtyr commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -0,0 +1,274 @@
+    output = flatpak_command(command)
+
+    if 'error' in output and 'already installed' not in output:
+        return 1, output
```
I kind of don't like returning from a function from a condition. What about to define `code` variable which you assign in the condition and then return at the end of the function?

```
    ...
    output = flatpak_command(command)
    code = 0

    if 'error' in output and 'already installed' not in output:
        code = 1

    return code, output
```

The same anywhere bellow.

JayKayy commented:
Great feedback and thank you for the formatting/syntactical help. Any more feedback?

sysadmind reviewed (approved):
Thanks for all of the updates!

JayKayy commented:
Are there any more TODO's on this?

jctanner commented:
@JayKayy you still have a merge commit that can only be resolved by following the rebase instructions posted by the bot previously.

https://github.com/ansible/ansible/commit/55e8c16ff71107f64c791f5960b228a00204cf16 --> "2 parents 52e8d92 + e419a55"

Until you resolve that, we will not merge this PR.

mattclay commented:
This PR contains many changes unrelated to the module being added.

needs_rebase

meebey commented:
@JayKayy will you rebase and revive this PR later? Easiest would be to just add the module file on a fresh topic branch as I see there are some merge conflict markers committed

Never mind, just found the other PR where you continue: ansible/pull/28742
